backends/p4tools/testgen/targets/ebpf/backend/stf/stf.cpp

```diff
@@ -0,0 +1,296 @@
+#include <inja/inja.hpp>
+
+#include "backends/p4tools/common/lib/format_int.h"
+#include "backends/p4tools/common/lib/ir.h"
+#include "backends/p4tools/common/lib/trace_events.h"
+#include "backends/p4tools/common/lib/util.h"
+#include "gsl/gsl-lite.hpp"
+#include "ir/ir.h"
+#include "lib/gmputil.h"
+#include "lib/log.h"
+#include "nlohmann/json.hpp"
+
+#include "backends/p4tools/testgen/lib/exceptions.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+namespace EBPF {
+
+STF::STF(cstring testName, boost::optional<unsigned int> seed = boost::none) : TF(testName, seed) {
+    boost::filesystem::path testFile(testName + ".stf");
+    cstring testNameOnly(testFile.stem().c_str());
+}
+
+inja::json STF::getTrace(const TestSpec* testSpec) {
+    inja::json traceList = inja::json::array();
+    const auto* traces = testSpec->getTraces();
+    if ((traces != nullptr) && !traces->empty()) {
+        for (const auto& trace : *traces) {
+            std::stringstream ss;
+            ss << *trace;
+            traceList.push_back(ss.str());
+        }
+    }
+    return traceList;
+}
+
+inja::json STF::getControlPlane(const TestSpec* testSpec) {
+    inja::json controlPlaneJson = inja::json::object();
+
+    auto tables = testSpec->getTestObjectCategory("tables");
+    if (!tables.empty()) {
+        controlPlaneJson["tables"] = inja::json::array();
+    }
+    for (auto const& testObject : tables) {
+        inja::json tblJson;
+        tblJson["table_name"] = testObject.first.c_str();
+        const auto* const tblConfig = testObject.second->checkedTo<TableConfig>();
+        auto const* tblRules = tblConfig->getRules();
+        tblJson["rules"] = inja::json::array();
+        for (const auto& tblRule : *tblRules) {
+            inja::json rule;
+            auto const* matches = tblRule.getMatches();
+            auto const* actionCall = tblRule.getActionCall();
+            auto const* actionArgs = actionCall->getArgs();
+            rule["action_name"] = actionCall->getActionName().c_str();
+            auto j = getControlPlaneForTable(*matches, *actionArgs);
+            rule["rules"] = std::move(j);
+            rule["priority"] = tblRule.getPriority();
+            tblJson["rules"].push_back(rule);
+        }
+
+        controlPlaneJson["tables"].push_back(tblJson);
+    }
+
+    return controlPlaneJson;
+}
+
+inja::json STF::getControlPlaneForTable(const std::map<cstring, const FieldMatch>& matches,
+                                        const std::vector<ActionArg>& args) {
+    inja::json rulesJson;
+
+    rulesJson["matches"] = inja::json::array();
+    rulesJson["act_args"] = inja::json::array();
+    rulesJson["needs_priority"] = false;
+
+    for (auto const& match : matches) {
+        auto fieldName = match.first;
+        auto const& fieldMatch = match.second;
+
+        // Replace header stack indices hdr[<index>] with hdr$<index>.
+        // TODO: This is a limitation of the stf parser. We should fix this.
```

Comment:
The STF parser can not handle header stacks as key names. We have to rewrite them. This is a strange limitation we should fix at some point.